### PR TITLE
Adds the ability to customize the first join, join, and disconnect messages

### PR DESCRIPTION
**Pull request by Meronat**
Adds the ability to disable or enable the welcome message and set the server name yourself including color codes by an ampersand in the bedrock configuration file.

### CODE REVIEW

**Comment by viveleroi** [comment]
This is a good feature to have but we should allow someone to truly customize it. The entire welcome message should be configurable, not just the server name, so we'll have something like `Welcome %s tot the server!" (with color code deserialization support, etc). The same should be applied to the quit messages.

**Comment by Meronat** [comment]
Okay, I will make changes to reflect this, but how would I translate the %s into the player name?

Also since you mentioned quit messages would you like the plugin to override current join/quit messages and be configured just like this first join message?

**Comment by viveleroi** [comment]
`String.format`, you can see other examples in the Bedrock code.

**Comment by viveleroi** [comment]
The plugin should override vanilla/default messages, yes.

**Comment by Meronat** [comment]
@viveleroi How does this look?

File: src/main/java/com/helion3/bedrock/Configuration.java [hunk 1]
```diff
@@ -66,6 +66,36 @@ public Configuration(File defaultConfig, ConfigurationLoader<CommentedConfigurat
                 afkKick.setValue(120);
             }
 
+            ConfigurationNode firstJoinMessageEnabled = rootNode.getNode("joinLeaveMessages", "firstJoin", "enabled");
```
Comment (viveleroi) [inline src/main/java/com/helion3/bedrock/Configuration.java]: We can keep the root node as `messages`, there could eventually be more than just these three and it's a bit redundant, `joinLeaveMessages.join` etc.

File: src/main/java/com/helion3/bedrock/Configuration.java [hunk 2]
```diff
@@ -66,6 +66,36 @@ public Configuration(File defaultConfig, ConfigurationLoader<CommentedConfigurat
...
+            if (firstJoinMessageEnabled.isVirtual()) {
+                firstJoinMessageEnabled.setValue(true);
+            }
+
+            ConfigurationNode firstJoinMessage = rootNode.getNode("joinLeaveMessages", "firstJoin", "message");
+            if (firstJoinMessage.isVirtual()) {
+                firstJoinMessage.setValue("&6Welcome &b%s &6to the server!");
+            }
+
+            ConfigurationNode joinMessageEnabled = rootNode.getNode("joinLeaveMessages", "join", "overrideEnabled");
```
Comment (viveleroi) [inline src/main/java/com/helion3/bedrock/Configuration.java]: I think these can be `enabled`. That's consistent with other configs but also, if it's enabled, we're going to override the default, so we don't need to call it `overrideEnabled`

File: src/main/java/com/helion3/bedrock/listeners/DisconnectListener.java [hunk 3]
```diff
@@ -36,16 +39,28 @@ public void onPlayerQuit(final ClientConnectionEvent.Disconnect event) {
             return;
         }
 
+        Player player = event.getTargetEntity();
+
+        // Is disconnect message enabled?
+        if (Bedrock.getConfig().getNode("joinLeaveMessages", "leave", "overrideEnabled").getBoolean()) {
+            String serializedDisconnectMessage = Bedrock.getConfig().getNode("joinLeaveMessages", "leave", "message").getString();
+
+            Text disconnectMessage = TextSerializers.formattingCode('&').deserialize(String.format(serializedDisconnectMessage, player.getName()));
```
Comment (viveleroi) [inline src/main/java/com/helion3/bedrock/listeners/DisconnectListener.java]: We can cache this formatted message as a class variable so we don't need to deserialize it every time someone disconnects. The same applies for the other messages.

File: src/main/java/com/helion3/bedrock/listeners/JoinListener.java [hunk 4]
```diff
@@ -49,8 +49,26 @@ public void onPlayerJoin(final ClientConnectionEvent.Join event) {
...
 
-            MessageChannel.TO_ALL.send(Text.of(TextColors.GOLD, "Welcome ", TextColors.LIGHT_PURPLE,
-                player.getName(),TextColors.GOLD, " to DHMC!"));
+            // Is welcome message enabled?
+            if (Bedrock.getConfig().getNode("joinLeaveMessages", "firstJoin", "enabled").getBoolean()) {
+                String serializedFirstJoinMessage = Bedrock.getConfig().getNode("joinLeaveMessages", "firstJoin", "message").getString();
+
+                Text firstJoinMessage = TextSerializers.formattingCode('&').deserialize(String.format(serializedFirstJoinMessage, player.getName()));
+
+                MessageChannel.TO_ALL.send(Text.of(firstJoinMessage));
```
Comment (viveleroi) [inline src/main/java/com/helion3/bedrock/listeners/JoinListener.java]: `firstJoinMessage` is already a `Text` instance, so you can leave out `Text.of()`

File: src/main/java/com/helion3/bedrock/Configuration.java (hunk 2 above)
Comment (Meronat) [inline src/main/java/com/helion3/bedrock/Configuration.java]: Okay so, as an example the firstJoin message would have the configs messages.firstJoin.enabled and messages.firstJoin.message? Likewise for the other two as well?

Comment (viveleroi) [inline src/main/java/com/helion3/bedrock/Configuration.java]: That works.

File: src/main/java/com/helion3/bedrock/listeners/DisconnectListener.java (hunk 3 above)
Comment (Meronat) [inline src/main/java/com/helion3/bedrock/listeners/DisconnectListener.java]: I'm not exactly sure how/where I would go about that.

Comment (viveleroi) [inline src/main/java/com/helion3/bedrock/listeners/DisconnectListener.java]: You can move the variable to the top of the class block:

    private final Text disconnectMessage;

and inside the class constructor, set it:

```
public DisconnectListener() {
    String message = Bedrock.getConfig().getNode("joinLeaveMessages", "leave", "message").getString();
    disconnectMessage = TextSerializers.formattingCode('&').deserialize(message));
}
```

This way `disconnectMessage` is available each time the listener is called, but it's only deserialized once.

    event.setMessage(String.format(disconnectMessage, player.getName()));

**Comment by Meronat** [comment]
Okay I will make all of the changes suggested, thanks for your help, it means a lot.

File: src/main/java/com/helion3/bedrock/listeners/DisconnectListener.java (hunk 3 above)
Comment (Meronat) [inline src/main/java/com/helion3/bedrock/listeners/DisconnectListener.java]: So I am attempting to make this change, but this line is giving me an error in IntelliJ.

Line: `event.setMessage(String.format(disconnectMessage, player.getName()));`

Error: `cannot resolve method 'format(org.spongepowered.api.text.Text, java.lang.String)'`

It seems like I can only apply the player name before the deserialization takes place, which does not work if we deserialize in the constructor.

Comment (viveleroi) [inline src/main/java/com/helion3/bedrock/listeners/DisconnectListener.java]: I believe `Text` may have some placeholder support as of API 4, you might ask in spongedev. If not, I guess there's no convenient way to cache it.

**Comment by Meronat** [comment]
Don't have time to work on this right now, may revisit it in the future from scratch.

**Pull request closed without merging**